localstack/services/s3/provider_stream.py

```diff
@@ -0,0 +1,808 @@
+            raise NoSuchUpload()
+
+        key = moto_backend.upload_part(
+            bucket_name, upload_id, request.get("PartNumber"), request.get("Body")
+        )
+        response = UploadPartOutput(ETag=key.etag)
+
+        if key.checksum_algorithm is not None:
+            response[f"Checksum{key.checksum_algorithm.upper()}"] = key.checksum_value
+
+        if key.encryption is not None:
+            response["ServerSideEncryption"] = key.encryption
+            if key.encryption == "aws:kms" and key.kms_key_id is not None:
+                response["SSEKMSKeyId"] = key.encryption
+
+        if key.encryption == "aws:kms" and key.bucket_key_enabled is not None:
+            response["BucketKeyEnabled"] = key.bucket_key_enabled
+
+        return response
+
+
+class ChecksumInvalid(s3_exceptions.S3ClientError):
+    code = 400
+
+    def __init__(self, algorithm: str):
+        super().__init__(
+            "InvalidRequest",
+            f"Value for x-amz-checksum-{algorithm.lower()} header is invalid.",
+        )
+
+
+class PartialStream(RawIOBase):
+    def __init__(self, base_stream: IO[bytes], start_byte: int, end_byte: int):
+        super().__init__()
+        self._base_stream = base_stream
+        self._pos = start_byte
+        self._length = end_byte - start_byte + 1
+
+    def read(self, s: int = -1) -> bytes | None:
+        self._base_stream.seek(self._pos)
+
+        if not s or s < 0:
+            amount = self._length
+        else:
+            amount = min(self._length, s)
+
+        data = self._base_stream.read(amount)
+        if not data:
+            return b""
+        read_amount = len(data)
+        self._length -= read_amount
+        self._pos += read_amount
+
+        return data
+
+
+class StreamedFakeKey(s3_models.FakeKey):
+    def __init__(self, name: str, value: IO[bytes], *args, **kwargs):
+        # when we set the value to nothing to first initialize the key for `PutObject` until we pull all logic in the
+        # provider
+        if not value or isinstance(value, bytes):
+            value = BytesIO(value or b"")
+        super(StreamedFakeKey, self).__init__(name, value, *args, **kwargs)
+        self.is_latest = True
+
+    @property
+    def value(self) -> IO[bytes]:
+        self._value_buffer.seek(0)
+        return self._value_buffer
+
+    @value.setter
+    def value(self, new_value: IO[bytes] | SpooledTemporaryFile):
+        etag_empty = not self._etag or self._etag == "d41d8cd98f00b204e9800998ecf8427e"
+        # it could come from the already calculated and completed CompleteMultipartUpload
+        # in that case, set it directly as the buffer
+        # if the etag is not set, this is the result from CopyObject, in that case we should copy the underlying
+        # SpooledTemporaryFile
+        stream_id = id(self._value_buffer)
+        key_lock = S3_KEY_LOCKS[stream_id]
+        if self._etag and isinstance(new_value, SpooledTemporaryFile):
+            with key_lock:
```

Comment:
I think theoretically, this lock handling can become problematic:
- Thread A: Creates the lock by accessing the dict
- Thread B: Tries to acquire the lock, has to wait
- Thread A: Finishes, removes the local from the dictionary, thread B can acquire the lock
- Thread C: Tries to get the lock, nothing is in the dict (even through thread B is still running), gets a new lock, can acquire it

Could this be a problem?

Reply:
I think you're right, see https://github.com/localstack/localstack/pull/8473#discussion_r1246735523, I'll try to implement it differently without the need for a global locking system.